libsignal-protocol/src/crypto.rs

```diff
@@ -56,49 +324,225 @@ impl CryptoProvider {
     user_data: *mut c_void,
 ) -> c_int {
     assert!(!data.is_null());
     assert!(!user_data.is_null());
 
     let user_data = &*(user_data as *const State);
-    let buffer = slice::from_raw_parts_mut(data, len);
-    user_data.0.fill_random(buffer);
-    0
+    let buffer = slice::from_raw_parts(data, data_len);
+    user_data.0.sha512_digest_update(buffer).into_code()
+}
+
+unsafe extern "C" fn sha512_digest_final_func(
+    _digest_context: *mut c_void,
+    output: *mut *mut signal_buffer,
+    user_data: *mut c_void,
+) -> c_int {
+    // just to make sure that the c ffi gave us a valid buffer to write to.
+    assert!(!output.is_null());
+    assert!(!user_data.is_null());
+
+    let user_data = &*(user_data as *const State);
+    match user_data.0.sha512_digest_final() {
+        Ok(buf) => {
+            let buffer = Buffer::from(buf);
+            *output = buffer.into_raw();
+            sys::SG_SUCCESS as c_int
+        },
+        Err(e) => e.code(),
+    }
+}
+
+unsafe extern "C" fn sha512_digest_cleanup_func(
+    _digest_context: *mut c_void,
+    user_data: *mut c_void,
+) {
+    assert!(!user_data.is_null());
+
+    let user_data = &*(user_data as *const State);
+    user_data.0.sha512_digest_cleanup();
+}
+
+unsafe extern "C" fn encrypt_func(
+    output: *mut *mut signal_buffer,
+    cipher: c_int,
+    key: *const u8,
+    key_len: usize,
+    iv: *const u8,
+    iv_len: usize,
+    plaintext: *const u8,
+    plaintext_len: usize,
+    user_data: *mut c_void,
+) -> c_int {
+    internal_cipher(
+        CipherMode::Encrypt,
+        output,
+        cipher,
+        key,
+        key_len,
+        iv,
+        iv_len,
+        plaintext,
+        plaintext_len,
+        user_data,
+    )
+}
+
+unsafe extern "C" fn decrypt_func(
+    output: *mut *mut signal_buffer,
+    cipher: c_int,
+    key: *const u8,
+    key_len: usize,
+    iv: *const u8,
+    iv_len: usize,
+    ciphertext: *const u8,
+    ciphertext_len: usize,
+    user_data: *mut c_void,
+) -> c_int {
+    internal_cipher(
+        CipherMode::Decrypt,
+        output,
+        cipher,
+        key,
+        key_len,
+        iv,
+        iv_len,
+        ciphertext,
+        ciphertext_len,
+        user_data,
+    )
+}
+
+#[inline]
+unsafe extern "C" fn internal_cipher(
+    mode: CipherMode,
+    output: *mut *mut signal_buffer,
+    cipher: c_int,
+    key: *const u8,
+    key_len: usize,
+    iv: *const u8,
+    iv_len: usize,
+    data: *const u8,
+    data_len: usize,
+    user_data: *mut c_void,
+) -> c_int {
+    use self::CipherMode::*;
+    // just to make sure that the c ffi gave us a valid buffer to write to.
+    assert!(!output.is_null());
+    assert!(!user_data.is_null());
+    assert!(!key.is_null());
+    assert!(!iv.is_null());
+    assert!(!data.is_null());
+
+    let signal_cipher_type = match SignalCipherType::try_from(cipher) {
+        Ok(ty) => ty,
+        // return early of the function with invalid arg instead of unknown
```

Comment:
I think there might have been a typo here. Did you mean "return early from the function"? 

It's nice to see we're handling `try_from()` errors though :+1: